semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/ScalaSigAttribute.scala

```diff
@@ -0,0 +1,26 @@
+package scala.meta.internal.scalacp
+
+import scala.tools.asm.Attribute
+import scala.tools.asm.ClassReader
+import scala.tools.asm.Label
+import scala.tools.scalap.Main
+import scala.tools.scalap.scalax.rules.scalasig.ByteCode
+import scala.tools.scalap.scalax.rules.scalasig.ScalaSig
+import scala.tools.scalap.scalax.rules.scalasig.ScalaSigAttributeParsers
+
+sealed abstract class BaseScalaSigAttribute extends Attribute(Main.SCALA_SIG) {
+  override def read(
+      cr: ClassReader,
+      off: Int,
+      len: Int,
+      buf: Array[Char],
+      codeOff: Int,
+      labels: Array[Label]
+  ): Attribute = {
+    val bytecode = new ByteCode(cr.b, off, len)
+    val scalaSig = ScalaSigAttributeParsers.parse(bytecode)
+    ScalaSigAttribute(scalaSig)
+  }
+}
+final case class ScalaSigAttribute(scalaSig: ScalaSig) extends BaseScalaSigAttribute
+object ScalaSigAttribute extends BaseScalaSigAttribute
```

Comment:
This class/companion business still looks mysterious to me. If I understand correctly, it's a clever way to reuse some code by fusing multiple concepts into one entity?

If yes, how about having two different classes: 1) `final case class ScalaSigAttribute(scalaSig: ScalaSig) extends extends Attribute(Main.SCALA_SIG)`, 2) `object ScalaSigPrototype extends Attribute(Main.SCALA_SIG) { def read... }`?

Reply:
That simplified a lot, great suggestion 👍